### PR TITLE
PC-1121 Preventing Duplicate Referrals

### PR DESCRIPTION
**Pull request by samyou-softwire**
[Link to Jira ticket](https://beisdigital.atlassian.net/browse/PC-1121)

# Description

adds a new page to warn a user if a referral has already been submitted

this hooks in after the "epc-select" page, or "address-select" page, depending on what route the user sees

the page is reasonably standard, puts some booleans in context & renders them

the next page is set to the same page that would be served in the relevant flow

translations are pending

# Checklist

- [x] I have made any necessary updates to the documentation
- [x] I have checked there are no unnecessary IDE warnings in this PR
- [x] I have checked there are no unintentional line ending changes
- [x] I have added tests where applicable
- [x] If I have made any changes to the code, I have used the IDE auto-formatter on it
- [x] If I have made any changes to Python files (even if not changing any content strings), I have run `make extract-translations`
- [ ] If I have made any changes to website flow, I have updated the [Flow Miro Board](https://miro.com/app/board/uXjVK4AgQ3E=/)
- [x] If I have made any changes to website flow, I have checked forward and back behaviour is still consistent

# Screenshots
image

Screenshot_22-7-2024_112628_localhost

### CODE REVIEW

File: help_to_heat/frontdoor/interface.py [hunk 1]
```diff
@@ -164,6 +164,36 @@ def replace_in_session_data(self, session_data):
         return session_data
 
 
+class DuplicateReferralChecker:
```
Comment (jdgage) [inline help_to_heat/frontdoor/interface.py:167]: How much effort would it be to have this live in its own file? I think we should try to move away from having these massive 'everything' files.

File: help_to_heat/frontdoor/interface.py [hunk 2]
```diff
@@ -164,6 +164,36 @@ def replace_in_session_data(self, session_data):
...
+        self.session_id = session_id
+
+    def is_referral_a_duplicate(self):
+        session_data = api.session.get_session(self.session_id)
+        uprn = session_data.get("uprn")
+        return uprn and portal.models.Referral.objects.filter(data__uprn=uprn).exists()
+
+    def is_duplicate_referral_sent_to_same_energy_supplier(self):
+        if not self.is_referral_a_duplicate():
+            return False
```
Comment (jdgage) [inline help_to_heat/frontdoor/interface.py:178]: Should we throw here? Something's gone wrong if we're at this point and there's no duplicate

File: help_to_heat/frontdoor/interface.py [hunk 3]
```diff
@@ -164,6 +164,36 @@ def replace_in_session_data(self, session_data):
...
+        session_data = api.session.get_session(self.session_id)
+        uprn = session_data.get("uprn")
+        # needs to compare against the supplier that would've been saved
+        saved_supplier = SupplierConverter(self.session_id).get_supplier_on_success_page()
+        referral = portal.models.Referral.objects.get(data__uprn=uprn)
+        return referral.supplier.name == saved_supplier
+
+    def get_date_of_previous_referral(self):
+        if not self.is_referral_a_duplicate():
+            return None
```
Comment (jdgage) [inline help_to_heat/frontdoor/interface.py:189]: Likewise

File: help_to_heat/frontdoor/interface.py [hunk 4]
```diff
@@ -164,6 +164,36 @@ def replace_in_session_data(self, session_data):
...
+
+    def is_duplicate_referral_sent_to_same_energy_supplier(self):
+        if not self.is_referral_a_duplicate():
+            return False
+
+        session_data = api.session.get_session(self.session_id)
+        uprn = session_data.get("uprn")
+        # needs to compare against the supplier that would've been saved
+        saved_supplier = SupplierConverter(self.session_id).get_supplier_on_success_page()
+        referral = portal.models.Referral.objects.get(data__uprn=uprn)
```
Comment (jdgage) [inline help_to_heat/frontdoor/interface.py:184]: What if there's more than one?

File: help_to_heat/frontdoor/interface.py [hunk 5]
```diff
@@ -164,6 +164,36 @@ def replace_in_session_data(self, session_data):
...
 
 
+class DuplicateReferralChecker:
+    def __init__(self, session_id):
+        self.session_id = session_id
+
+    def is_referral_a_duplicate(self):
+        session_data = api.session.get_session(self.session_id)
+        uprn = session_data.get("uprn")
+        return uprn and portal.models.Referral.objects.filter(data__uprn=uprn).exists()
```
Comment (jdgage) [inline help_to_heat/frontdoor/interface.py:174]: There was some discussion of having a cutoff date - did that get explicitly discarded? We should probably check if not.

**Review by jdgage — CHANGES_REQUESTED** [review]

File: help_to_heat/frontdoor/interface.py (hunk 5 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/interface.py:174]: will ask on the ticket

File: help_to_heat/frontdoor/interface.py (hunk 4 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/interface.py:184]: have reworked the code a bit to account for this

File: help_to_heat/frontdoor/interface.py (hunk 2 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/interface.py:178]: sure

File: help_to_heat/frontdoor/interface.py (hunk 1 above)
Comment (jdgage) [reply help_to_heat/frontdoor/interface.py:171]: I think that's fine, but interested in what specifically?

Comment (samyou-softwire) [reply help_to_heat/frontdoor/interface.py:171]: SupplierConverter references api (in interface.py)
interface.py references SupplierConverter (in Session entity)

could leave SupplierConverter where it is?

Comment (jdgage) [reply help_to_heat/frontdoor/interface.py:171]: Yeah I'm happy with that for what it's worth (I think trying to restructure will quickly become a big job as we pull on the strings, for reasons like this - so we're best off just trying to do right with the new things we add for now)

File: help_to_heat/frontdoor/interface.py [hunk 6]
```diff
@@ -164,33 +164,42 @@ def replace_in_session_data(self, session_data):
...
+class NoMatchingReferralInSessionException(Exception):
+    pass
+
+
 class DuplicateReferralChecker:
     def __init__(self, session_id):
         self.session_id = session_id
 
-    def is_referral_a_duplicate(self):
+    def _try_find_duplicate_referral(self):
```
Comment (jdgage) [inline help_to_heat/frontdoor/interface.py:175]:
```suggestion
    def _try_find_most_recent_duplicate_referral(self):
```
perhaps?

**Review by jdgage — CHANGES_REQUESTED** [review]

File: help_to_heat/frontdoor/interface.py (hunk 6 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/interface.py]: sure

File: help_to_heat/frontdoor/views.py [hunk 7]
```diff
@@ -14,7 +14,9 @@
 from ..portal import email_handler
 from . import eligibility, interface, schemas
 from .eligibility import calculate_eligibility, eco4
-from .interface import DuplicateReferralChecker
+from .session_handlers.duplicate_referral_checker import (
+    DuplicateReferralChecker,
+)
```
Comment (jdgage) [inline help_to_heat/frontdoor/views.py:19]: Is this a formatting thing for line length?

**Review by jdgage — APPROVED** [review]
All looks fine I think. To check, as I have it we're blocked on:
1. A decision about the exit button
2. Translations
3. The cutoff date question

Is that right?

File: help_to_heat/frontdoor/views.py (hunk 7 above)
Comment (samyou-softwire) [reply help_to_heat/frontdoor/views.py:19]: yeah, very conservative but it insists

File: help_to_heat/templates/frontdoor/referral-already-submitted.html [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+            <input class="govuk-checkboxes__input" id="question-submit-another" name="submit_another" type="checkbox" {% if data.get('submit_another') == 'on' %}checked{% endif %}/>
+            <label class="govuk-label govuk-checkboxes__label" for="question-submit-another" id="question-submit-another-label">
+              {{_("I want to submit another referral")}}
+            </label>
+          </div>
+        </div>
+      </fieldset>
+    </div>
+  {%- endcall %}
+  <a href="https://www.gov.uk/apply-great-british-insulation-scheme" role="button" draggable="false" class="govuk-button" data-module="govuk-button">Exit service</a>
```
Comment (jdgage) [inline help_to_heat/templates/frontdoor/referral-already-submitted.html:37]: @Glenn-Clarke if you get around to picking this up, would you mind checking that this is consistent with what we do on RTV? It feels like the most sensible thing to me, but want to check!

Comment (Glenn-Clarke) [reply help_to_heat/templates/frontdoor/referral-already-submitted.html:37]: yep, I'll give it a look. I believe during the handover yesterday Sammy said this was how they did it, but I'll double check as I might have misremembered.

Comment (jdgage) [reply help_to_heat/templates/frontdoor/referral-already-submitted.html:37]: As discussed, let's change to just a link to the `gov.uk` homepage

File: help_to_heat/frontdoor/views.py [hunk 9]
```diff
@@ -487,7 +487,9 @@ def save_data(self, request, session_id, page_name, *args, **kwargs):
         return data
 
     def handle_post(self, request, session_id, page_name, data, is_change_page):
-        if DuplicateReferralChecker(session_id).is_referral_a_duplicate():
+        duplicate_referral_checker = DuplicateReferralChecker(session_id)
+        if (duplicate_referral_checker.is_referral_a_duplicate()
+                and duplicate_referral_checker.is_duplicate_referral_recent()):
```
[the remark this answers is not in the document]
Comment (Glenn-Clarke) [reply help_to_heat/frontdoor/views.py:492]: Good idea! I've just been taught about Django query filters (Thanks Sammy 😄)

File: tests/utils.py [hunk 10]
```diff
@@ -128,3 +129,50 @@ def login(client, email, password):
...
+            "user_selected_supplier": "British Gas",
+            "building_name_or_number": "10",
+            "property_main_heat_source": "",
+        }
+    supplier = portal.models.Supplier.objects.get(name=data["supplier"])
+    referral = portal.models.Referral.objects.create(session_id=session_id, data=data, supplier=supplier)
+    referral.save()
+    referral = portal.models.Referral.objects.get()
+    referral.created_at = creation_timestamp
+    referral.save()
```
Comment (jdgage) [inline tests/utils.py:178]: This is presumably because it autofills `created_at` when first creating?

File: tests/test_frontdoor.py [hunk 11]
```diff
@@ -269,7 +272,7 @@ def _do_happy_flow(supplier="EON", benefits_answer="Yes", park_home=False, house
         supplier_shown = "E.ON Next"
     if supplier == "Shell":
         supplier_shown = "Octopus Energy"
-
+    print(page.path)
```
Comment (jdgage) [inline tests/test_frontdoor.py]:
```suggestion
```

**Review by jdgage — CHANGES_REQUESTED** [review]

File: tests/utils.py (hunk 10 above)
Comment (Glenn-Clarke) [reply tests/utils.py:178]: That's correct!

**Review by jdgage — APPROVED** [review]

File: tests/test_frontdoor.py (hunk 11 above)
Comment (Glenn-Clarke) [reply tests/test_frontdoor.py]: This change should already be in! I ran check-code locally and it didn't flag any changes to test_frontdoor for some reason. I've made the changes it asked for now

**Pull request merged**